### PR TITLE
BUG: Fix nested_to_record with None values in nested levels

### LINKED ISSUE #21158: json_normalize gives KeyError in 0.23
#### Code Sample, a copy-pastable example if possible

```python
import json
from pandas import show_versions
from pandas.io.json import json_normalize

print(show_versions())

with open('test.json', 'r') as infile:
    d = json.load(infile)

normed = json_normalize(d)
```

The `test.json` file is rather lengthy, with a structure similar to:

```json
{
  "subject": {
    "pairs": {
      "A1-A2": {
        "atlases": {
          "avg.corrected": {
            "region": null,
            "x": 49.151580810546875,
            "y": -33.148521423339844,
            "z": 27.572303771972656
          }
        }
      }
    }
  }
}
```

This minimal version is enough to show the error below.

#### Problem description

This problem is *new* in pandas 0.23. I get the following traceback:

Traceback:

```pytb
Traceback (most recent call last):
  File "test.py", line 10, in <module>
    normed = json_normalize(d)
  File "/Users/depalati/miniconda3/lib/python3.6/site-packages/pandas/io/json/normalize.py", line 203, in json_normalize
    data = nested_to_record(data, sep=sep)
  File "/Users/depalati/miniconda3/lib/python3.6/site-packages/pandas/io/json/normalize.py", line 88, in nested_to_record
    new_d.update(nested_to_record(v, newkey, sep, level + 1))
  File "/Users/depalati/miniconda3/lib/python3.6/site-packages/pandas/io/json/normalize.py", line 88, in nested_to_record
    new_d.update(nested_to_record(v, newkey, sep, level + 1))
  File "/Users/depalati/miniconda3/lib/python3.6/site-packages/pandas/io/json/normalize.py", line 88, in nested_to_record
    new_d.update(nested_to_record(v, newkey, sep, level + 1))
  File "/Users/depalati/miniconda3/lib/python3.6/site-packages/pandas/io/json/normalize.py", line 84, in nested_to_record
    new_d.pop(k)
KeyError: 'region'
```

Note that running the same code on pandas 0.22 does not result in any errors. I suspect this could be related to #20399.

#### Expected Output

Expected output is a flattened `DataFrame` without any errors.

#### Output of ``pd.show_versions()``

<details>

INSTALLED VERSIONS
------------------
commit: None
python: 3.6.4.final.0
python-bits: 64
OS: Darwin
OS-release: 16.7.0
machine: x86_64
processor: i386
byteorder: little
LC_ALL: None
LANG: en_US.UTF-8
LOCALE: en_US.UTF-8

pandas: 0.23.0
pytest: 3.5.1
pip: 9.0.1
setuptools: 38.4.0
Cython: None
numpy: 1.14.2
scipy: None
pyarrow: None
xarray: 0.10.3
IPython: 6.3.1
sphinx: 1.7.2
patsy: None
dateutil: 2.7.2
pytz: 2018.3
blosc: None
bottleneck: None
tables: 3.4.2
numexpr: 2.6.4
feather: None
matplotlib: 2.2.2
openpyxl: None
xlrd: None
xlwt: None
xlsxwriter: None
lxml: None
bs4: 4.6.0
html5lib: 1.0.1
sqlalchemy: 1.2.7
pymysql: None
psycopg2: None
jinja2: 2.10
s3fs: None
fastparquet: None
pandas_gbq: None
pandas_datareader: None

</details>

### PR DESCRIPTION
**Pull request by ssikdar1**
- [ ] closes #21158
- [ ] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

Continue after pop so you dont pop with the same key twice in a row

### CODE REVIEW

**Comment by ssikdar1** [comment]
>>  git diff upstream/master -u -- "*.py" | flake8 --diff
$ git diff upstream/master -u -- "*.py" | flake8 --diff
$

File: pandas/io/json/normalize.py [hunk 1]
```diff
@@ -80,6 +80,7 @@ def nested_to_record(ds, prefix="", sep=".", level=0):
                 if level != 0:  # so we skip copying for top level, common case
                     v = new_d.pop(k)
                     new_d[newkey] = v
+                    continue
                 if v is None:  # pop the key if the value is None
```
Comment (WillAyd) [inline pandas/io/json/normalize.py]: Perhaps instead of continue this should just be `elif`?

**Review by WillAyd — CHANGES_REQUESTED** [review]
Thanks for the change - you can add a whatsnew for v0.23.1

**Comment by ssikdar1** [comment]
can you take a look at my newest commit?

**Comment by ssikdar1** [comment]
Can you check my latest commits? I changed it to not use a continue.

Also changed the whatsnew as well.

File: pandas/io/json/normalize.py [hunk 2]
```diff
@@ -80,8 +80,9 @@ def nested_to_record(ds, prefix="", sep=".", level=0):
                 if level != 0:  # so we skip copying for top level, common case
                     v = new_d.pop(k)
                     new_d[newkey] = v
-                if v is None:  # pop the key if the value is None
-                    new_d.pop(k)
+                else:
```
Comment (WillAyd) [inline pandas/io/json/normalize.py]: Use elif instead of a separate else and if

**Review by WillAyd — CHANGES_REQUESTED** [review]

**Comment by ssikdar1** [comment]
kk, changed to elif

File: doc/source/whatsnew/v0.23.1.txt [hunk 3]
```diff
@@ -45,6 +45,7 @@ Bug Fixes
 ~~~~~~~~~
 
 - tab completion on :class:`Index` in IPython no longer outputs deprecation warnings (:issue:`21125`)
+- fix normalize to not except when json_normalize is fed a key with null value on inner most key (:issue:`21158`)
```
Comment (jreback) [inline doc/source/whatsnew/v0.23.1.txt]: move to IO section, use a ref to ``json_normalize``.  see if you can describe this is more plain english.

**Review by jreback — CHANGES_REQUESTED** [review]

**Comment by ssikdar1** [comment]
Moved into the I/O section in whatsnew. Changed the message hopefully its clearer.

File: doc/source/whatsnew/v0.23.1.txt [hunk 4]
```diff
@@ -81,7 +81,7 @@ Indexing
 I/O
 ^^^
 
--
+- bug in :meth:`nested_to_record` when :meth:`json_normalize` was called with certain nested jsons  (:issue:`21158`)
```
Comment (WillAyd) [inline doc/source/whatsnew/v0.23.1.txt]: Perhaps better said "Bug in ... with None values in nested levels"

File: pandas/tests/io/json/test_normalize.py [hunk 5]
```diff
@@ -375,3 +375,27 @@ def test_nonetype_dropping(self):
              'info.last_updated': '26/05/2012'}]
 
         assert result == expected
+
+    def test_nonetype_inner_most_level(self):
```
Comment (WillAyd) [inline pandas/tests/io/json/test_normalize.py]: Can you check if we have a test case to handle None values that show up in the initial level?

**Review by WillAyd — CHANGES_REQUESTED** [review]

**Comment by ssikdar1** [comment]
@WillAyd 

>> Can you check if we have a test case to handle None values that show up in the initial level?

These test case below have a json list with a key as None in the initial level of a json.

https://github.com/pandas-dev/pandas/blob/master/pandas/tests/io/json/test_normalize.py#L354

  data = [
            {'info': None,
             'author_name':
             {'first': 'Smith', 'last_name': 'Appleseed'}
             },
            {'info':
                {'created_at': '11/08/1993', 'last_updated': '26/05/2012'},
             'author_name':
                {'first': 'Jane', 'last_name': 'Doe'}
             }
        ]

https://github.com/pandas-dev/pandas/blob/master/pandas/tests/io/json/test_normalize.py#L58

[
        {'info': None},
        {'info':
            {'created ...

File: doc/source/whatsnew/v0.23.1.txt [hunk 6]
```diff
@@ -81,7 +81,7 @@ Indexing
 I/O
 ^^^
 
--
+- bug in :meth:`nested_to_record` when :meth:`json_normalize` was called with None values in nested levels  (:issue:`21158`)
```
Comment (jreback) [inline doc/source/whatsnew/v0.23.1.txt]: nested_to_record is not public, don't use it here

use double backticks around None

say nested levels in JSON

**Review by jreback — CHANGES_REQUESTED** [review]

File: doc/source/whatsnew/v0.23.1.txt [hunk 7]
```diff
@@ -81,7 +81,7 @@ Indexing
 I/O
 ^^^
 
--
+- bug when :meth:`json_normalize` was called with `None` values in nested levels in JSON  (:issue:`21158`)
```
Comment (WillAyd) [inline doc/source/whatsnew/v0.23.1.txt]: Capitalize "Bug" and as Jeff mentioned use double backticks around None, not just single

File: pandas/tests/io/json/test_normalize.py (hunk 5 above)
Comment (WillAyd) [reply pandas/tests/io/json/test_normalize.py]: Thanks for checking that we had a test to cover None at the top-most level. I suppose it makes sense to test None at the deepest level (which you have) and also somewhere in the middle, to make sure it doesn't mess up subsequent un-nesting operations.

Can you at least add a test with None in one of the middle layers? You can probably simplify the example below (doesn't need to be a direct copy/paste from the issue) and use parametrization to re-use the logic for different inputs (i.e. one with None at the deepest level, one with None somewhere in the middle).  Depending on how that looks maybe you want to even move the example with None at the top level here as well

**Review by WillAyd — CHANGES_REQUESTED** [review]

**Comment by ssikdar1** [comment]
>> Can you at least add a test with None in one of the middle layers? 
I changed the test to be a test with None at multiple levels.

+        data = {
 +            "id": None,
 +            "location": {
 +                "id": None,
 +                "country": {
 +                    "id": None,
 +                    "state": {
 +                        "id": None,
 +                        "town.info": {
 +                            "region": None,
 +                            "x": 49.151580810546875,
 +                            "y": -33.148521423339844,
 +                            "z": 27.572303771972656}}}
 +            }
 +        }

Capitalized the b in Bug and put None in ticks

File: doc/source/whatsnew/v0.23.1.txt [hunk 8]
```diff
@@ -81,7 +81,7 @@ Indexing
 I/O
 ^^^
 
--
+- Bug when :meth:`json_normalize` was called with `None` values in nested levels in JSON  (:issue:`21158`)
```
Comment (jreback) [inline doc/source/whatsnew/v0.23.1.txt]: this wont' render, you need pandas.io.json.json_normalize I think

Comment (jreback) [reply doc/source/whatsnew/v0.23.1.txt]: use double back ticks around ``None``

File: pandas/tests/io/json/test_normalize.py [hunk 9]
```diff
@@ -375,3 +375,32 @@ def test_nonetype_dropping(self):
              'info.last_updated': '26/05/2012'}]
 
         assert result == expected
+
+    def test_nonetype_multiple_levels(self):
+        # GH21158: If inner level json has a key with a null value
+        # make sure it doesnt do a new_d.pop twice and except
+        data = {
+            "id": None,
```
Comment (jreback) [inline pandas/tests/io/json/test_normalize.py]: does this have the same result if id is NOT repeated (with None), just missing at various levels? e.g. try with a single id with None at the top and bottom levels (in another case, leave this one as well)

**Review by jreback — CHANGES_REQUESTED** [review]

**Comment by jreback** [comment]
pls rebase as well

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
looks ok to me. @WillAyd pls merge when you are satisifed.

**Comment by WillAyd** [comment]
@jreback somewhat conflicted on this. This PR certainly fixes the issue at hand, but what is the point of us dropping `None` values at the top level of the JSON yet maintaining it at deeper levels? Seems like it would be more consistent from an end user perspective to maintain at all levels and leave it to them to explicitly drop if they wanted.

Could probably go forward with this for 0.23.1 as the bug fix but then revisit top level `None`-handling for 0.24.0?

**Comment by jreback** [comment]
@WillAyd hmm, that's a good point. shouldn't ``id`` be in the expected? why are we dropping it at all? yeah that sees wrong

**Comment by WillAyd** [comment]
As @ssikdar1 points out it does stem from #20030, though in reading through that and the subsequent PR I think it is confusing what was actually being addressed. The PR would imply that automatically dropping the top-level `None` value is a feature, though I would argue that's not desirable and is in fact a bug.

**Comment by jreback** [comment]
@ssikdar1 can you rebase

@WillAyd ok with this?

**Comment by WillAyd** [comment]
Yep we can go with this for now. I'll open a separate issue for the top-level None parsing

**Comment by WillAyd** [comment]
Thanks @ssikdar1 !

**Comment by ssikdar1** [comment]
No problem!

**Pull request merged**